sycl/source/device_selector.cpp

```diff
@@ -50,36 +52,50 @@ static int getDevicePreference(const device &Device) {
   return Score;
 }
 
-device device_selector::select_device() const {
-  std::vector<device> devices = device::get_devices();
-  int score = REJECT_DEVICE_SCORE;
+static void traceDeviceSelection(const device &Device, int Score, bool Chosen) {
+  bool shouldTrace = false;
+  if (Chosen) {
+    shouldTrace = detail::pi::trace(detail::pi::TraceLevel::PI_TRACE_BASIC);
+  } else {
+    shouldTrace = detail::pi::trace(detail::pi::TraceLevel::PI_TRACE_ALL);
+  }
+  if (shouldTrace) {
+    std::string PlatformName = Device.get_info<info::device::platform>()
+                                   .get_info<info::platform::name>();
+    std::string DeviceName = Device.get_info<info::device::name>();
+    auto selectionMsg = Chosen ? "Selected device: -> final score = "
+                               : "Candidate device: -> score = ";
+
+    std::cout << "SYCL_PI_TRACE[all]: " << selectionMsg << Score
+              << ((Score < 0) ? " (REJECTED)" : "") << std::endl
+              << "SYCL_PI_TRACE[all]: "
+              << "  platform: " << PlatformName << std::endl
+              << "SYCL_PI_TRACE[all]: "
+              << "  device: " << DeviceName << std::endl;
+  }
+}
+
+device select_device(DSelectorInvocableType DeviceSelectorInvocable,
+                     std::vector<device> &Devices) {
+  int score = detail::REJECT_DEVICE_SCORE;
   const device *res = nullptr;
 
-  for (const auto &dev : devices) {
-    int dev_score = (*this)(dev);
-
-    if (detail::pi::trace(detail::pi::TraceLevel::PI_TRACE_ALL)) {
-      std::string PlatformName = dev.get_info<info::device::platform>()
-                                     .get_info<info::platform::name>();
-      std::string DeviceName = dev.get_info<info::device::name>();
-      std::cout << "SYCL_PI_TRACE[all]: "
-                << "select_device(): -> score = " << dev_score
-                << ((dev_score < 0) ? " (REJECTED)" : "") << std::endl
-                << "SYCL_PI_TRACE[all]: "
-                << "  platform: " << PlatformName << std::endl
-                << "SYCL_PI_TRACE[all]: "
-                << "  device: " << DeviceName << std::endl;
-    }
+  for (const auto &dev : Devices) {
+    int dev_score = std::invoke(DeviceSelectorInvocable, dev);
```

Comment:
```suggestion
    int dev_score = DeviceSelectorInvocable(dev);
```
?

Reply:
@keryell   Since we are using `std::is_invocable_r_v`  when we screen it, I actually prefer the symmetry of using `std::invoke` to invoke it.  Maybe I'm in the minority on that. Would you mind if we kept it as is?

Reply:
Are you kidding me or are you a member of the Church of the C++ Discouragement and Obfuscation? :-)
Why not rewriting all the code with *Operator function objects* from https://en.cppreference.com/w/cpp/utility/functional instead of using just `+` or `-`? ;-)

Reply:
I don't want to start a holy war. I'll change it.  
The hallowed symmetrical beauty of std::is_invocable to std::invoke is perhaps not cherished by the unenlightened.